Vald architecture document (#366)
vdaas/vald: docs/user/architecture.md

    - [Agent Discoverer](#agent-discoverer)
  - [Vald Core Engine](#vald-core-engine)
    - [Vald Agent](#vald-agent)
    - [Vald Agent Scheduler](#vald-agent-scheduler)
    - [Vald Index Manager](#vald-index-manager)
  - [Vald Replication Manager](#vald-replication-manager)
    - [Vald Replication Manager Agent](#vald-replication-manager-agent)
    - [Vald Replication Manager Controller](#vald-replication-manager-controller)
  - [Kubernetes Components](#kubernetes-components)
    - [Kube-apiserver](#kube-apiserver)
    - [Custom Resources](#custom-resources)

## Overview

This document describes the high-level architecture design of Vald and explains each component in Vald. We need these components to support scalability, high performance, and auto-healing in Vald.

<img src="../../design/Vald Future Architecture Overview.svg" />

Vald is based on [Kubernetes](https://kubernetes.io/) architecture. Before you read this document, you must understand the basic concept of Kubernetes.

## Data Flow

In this section, we will explain the data flow in Vald which is the most important for users.

### Insert

<img src="./insert_search_flow.svg" />

When the user inserts data into Vald:

> ```

Reviewer: ```suggestion
```bash
```

> C1. Vald Ingress receives the request from the user. The request includes the vector and the vector ID.

Reviewer: ```suggestion
C1. Vald Ingress receives the request from the user. The request includes the vector and the vector ID which is set by user.
```

C2. Vald Ingress will forward the request to Vald Meta Gateway. After the Vald Meta Gateway receive the request, the UUID will be generated for internal use.
C3. Vald Meta Gateway will forward the request with UUID to Vald Backup Gateway.
C4. Vald Backup Gateway will forward the request to Vald LB Gateway.

> C5. Vald LB Gateway will decide which Vald Agent instance to process the request base on the node resource usage and forwards the request to the decided Vald Agent.

Reviewer: ```suggestion
C5. Vald LB Gateway will decide which Vald Agent instance to process the request based on the node resource usage and forward the request to the decided Vald Agent.
```

C6. Vald Agent returns success and the corresponding Vald Agent IP address to Vald LB Gateway.

> C7. Vald LB Gateway return success and the IP address to Vald Backup Gateway.

Reviewer: ```suggestion
C7. Vald LB Gateway returns success and the IP address to Vald Backup Gateway.
```

S12. Vald Meta Gateway will forward the searching result to the Vald Meta.
S13. Vald Meta will perform a search for the Vector ID base on the UUID and return the Vector ID to the Vald Meta Gateway.
S14. Vald Meta Gateway returns the searching result with the vector ID to the Vald Filter Gateway.
S15. Vald Filter Gateway will forward the request to Vald Egress Gateway to filter the final result.
```

<!-- ### Update -->

<!-- ### Delete -->

## Components

### Vald Filter

Vald Filter is an optional functionality in Vald.
User can implement the custom filtering logic and integrate with Vald.

Vald Filter provides the following functionalities.

- Custom filter base on request query
- Custom filter for the searching result

#### Vald Ingress Filter

Vald Ingress Filter filters the incoming request before processing it.

Users can implement custom filtering logic such as changing the vectors or filtering based on user ID.

#### Vald Egress Filter

Vald Egress Filter filters the response before sending it to the user.

This component can reorder the searching result from multiple Vald Agents based on the user-defined ranking.

#### Vald Filter Gateway

> Vald Filter Gateway forwards the request to Vald Ingress Filter before processing it and forwards the request to the Vald Egress Filter before returning the searching result to the user.

Reviewer: ```suggestion
Vald Filter Gateway forwards the request to Vald Ingress Filter before processing it and forwards the response to the Vald Egress Filter before returning the searching result to the user.
```
`response` is right?

### Vald Metadata

> In Vald, metadata is the vector data and the corresponding additional data to represent the set of the searching criteria and the result.

Reviewer: ```suggestion
In Vald, metadata consists of the vector data and the corresponding additional data to represent the set of the searching criteria and the result.
```

Each Vald Agent pod holds different high dimensional vector data space, which is constructed by insert/update vectors for searching approximate vectors.

When you request searching with your vector in Vald, each Vald Agent returns different _k_-nearest neighbors' vectors which are similar to the searching vector.

<img src="../../assets/doc/vector_data_space_explain.svg" />

The same vector will be inserted into multiple Vald Agents, so the vector data space constructed in each Vald Agent will be different, causing the searching results from each Vald Agent are different.

#### Vald Agent Scheduler

Vald Agent Scheduler is the scheduler of the Vald Agent.
It implements it's own custom scheduling logic to increase the scalability of the Vald Agent.

It schedules Vald Agent base on the Node CPU and memory usage, and the amount of the indexes.

#### Vald Index Manager

Vald Index Manager controls the timing of the indexing inserted vectors on the Vald Agent. The index is used to increase the performance of the search action.

It retrieves the active Vald Agent pods from the Vald Discoverer and triggers the indexing action on each Vald Agent.

### Vald Replication Manager

> Vald Replication Manager manages the healthiness of the Vald Agent. When the pod is dead, Vald Replication Manager will auto recover the cache to keeps the reliability of the service.

Author: ```suggestion
Vald Replication Manager manages the healthiness of the Vald Agent. 
When the pod is dead, Vald Replication Manager will recover the cache automatically to keeps the reliability of the service.
```

### Search

S1. Vald Ingress receives the search request from the user. The request includes the vector or the vector ID.
S2. Vald Ingress will forward the request to Vald Filter Gateway.
S3. Vald Filter Gateway will forward the request to Vald Ingress Filter.
S4. Vald Ingress Filter will perform the filtering and return the filtering result to the Vald Filter Gateway.
S5. Vald Filter Gateway will forward the request to the Vald Meta Gateway.
S6. Vald Meta Gateway will forward the request to the Vald Backup Gateway.
S7. Vald Backup Gateway will forward the request to the Vald LB Gateway.

> S8. Vald LB Gateway will decide which Vald Agent instance to process the request base on the node resource usage and forwards the request to the decided Vald Agent.

Reviewer 2: ```suggestion
S8. Vald LB Gateway will decide which Vald Agent instance to process the request base on the node resource usage and forward the request to the decided Vald Agent.
```
